api/opds.py

```diff
@@ -628,6 +641,15 @@ def __init__(
         self.identifies_patrons = library_identifies_patrons
         self.facets = facets or None
 
+        # Open access links are added only when the library needs no authentication,
+        # else we must not have direct fulfillment links in the feed.
+        # The Patron should go through the authenticated borrow workflow in that case.
+        _db = Session.object_session(self.library)
+        integrations = ExternalIntegration.for_library_and_goal(
+            _db, self.library, ExternalIntegration.PATRON_AUTH_GOAL
+        )
+        self.add_open_access_links = integrations.count() == 0
```

Comment:
Should this also depend on `api.can_fulfill_without_loan` ?

Reply:
Hmm... Thanks for pointing out locations in code. After looking at them it seems only `OPDSForDistributorsAPI` implements this method, whereas other datasource APIs simply `return False` (from the Base API).
That would defeat the purpose here since we're looking at open-access links from all sources. I won't use it here.

Reply:
Not sure. The `circulation` implementation in particular would seem to return `True` here:

```
def can_fulfill_without_loan(self, patron, pool, lpdm):
        """Can we deliver the given book in the given format to the given
        patron, even though the patron has no active loan for that
        book?

        In general this is not possible, but there are some
        exceptions, managed in subclasses of BaseCirculationAPI.

        :param patron: A Patron. This is probably None, indicating
            that someone is trying to fulfill a book without identifying
            themselves.

        :param delivery_mechanism: The LicensePoolDeliveryMechanism
            representing a format for a specific title.
        """
        if not lpdm or not pool:
            return False
        if pool.open_access:
            return True
        api = self.api_for_license_pool(pool)
        if not api:
            return False
        return api.can_fulfill_without_loan(patron, pool, lpdm)
```

Should there be an open access pool available?

Reply:
Oh :sweat_smile: In which case it's redundant since we only use it for open access pools :laughing:

Reply:
@RishiDiwanTT instead of querying for integration here and creating a new variable for it or using `can_fulfill_without_loan`, I think we can use `self.identifies_patrons`. If we identify patrons, we shouldn't include open access acquisition links in the feed.

Reply:
This variable depends on the `IDENTIFIES_INDIVIDUALS` class variable [[source](https://github.com/ThePalaceProject/circulation/blob/main/api/authenticator.py#L727)]. It looks like all AuthenticationProviders have `IDENTIFY_INDIVIDUALS=True` due to the [base class](https://github.com/ThePalaceProject/circulation/blob/main/api/authenticator.py#L1302).
Would it be equivalent to not having authentication (in this case only) if an authenticator is introduced that does not identify individuals? It seems this might make sense since there probably would be no "login" mechanism to the authenticator.

Reply:
@RishiDiwanTT The code is a bit of a maze, but I think (`LibraryAnnotator.library_identifies_patrons`)[https://github.com/ThePalaceProject/circulation/blob/3894035d6231685128cff97f9163d8fff2786cf9/api/opds.py#L608] is set when the class is instantiated. 

`LibraryAnnotator` gets instantiated by `CirculationManager` which sets it based on `library_identifies_patrons`:
https://github.com/ThePalaceProject/circulation/blob/1c565ad26555dec72cff819149c3004443e05a47/api/controller.py#L578-L581

So using `self.library_identifies_patrons` here, should cover two cases: 
- We are using an authentication method that doesn't identify individuals (which there are none)
- No authentication is configured

It seems like using this existing variable is a better choice then doing a new query here.

Reply:
Sounds good, will make the change with the assumption IDENTIFY_INDIVIDUALS will not be False anytime soon.

Reply:
I think if `IDENTIFY_INDIVIDUALS` does become false in the future, this will give us the behaviour we want.